src/platform/datapath_raw_socket.c

```diff
@@ -320,14 +332,90 @@ CxPlatFramingWriteHeaders(
+    _In_ QUIC_ADDRESS_FAMILY Family
     )
 {
     UDP_HEADER* UDP = (UDP_HEADER*)(Buffer->Buffer - sizeof(UDP_HEADER));
-    IPV4_HEADER* IP = (IPV4_HEADER*)(((uint8_t*)UDP) - sizeof(IPV4_HEADER));
-    ETHERNET_HEADER* Ethernet = (ETHERNET_HEADER*)(((uint8_t*)IP) - sizeof(ETHERNET_HEADER));
+    ETHERNET_HEADER* Ethernet;
+    IPV4_HEADER* IPv4;
+    IPV6_HEADER* IPv6;
+    uint16_t EthType;
+    uint16_t IpHeaderLen;
+
+    CXPLAT_DBG_ASSERT(
+        Family == QUIC_ADDRESS_FAMILY_INET || Family == QUIC_ADDRESS_FAMILY_INET6);
+
+    //
+    // Fill UDP header.
+    //
+    UDP->DestinationPort = RemoteAddress->Ipv4.sin_port;
+    UDP->SourcePort = LocalAddress->Ipv4.sin_port;
+    UDP->Length = QuicNetByteSwapShort((uint16_t)Buffer->Length);
+
+    //
+    // Fill IPv4/IPv6 header.
+    //
+    if (Family == QUIC_ADDRESS_FAMILY_INET) {
+        IPv4 = (IPV4_HEADER*)(((uint8_t*)UDP) - sizeof(IPV4_HEADER));
+        IPv4->VersionAndHeaderLength = IPV4_DEFAULT_VERHLEN;
+        IPv4->TypeOfService = 0;
+        IPv4->TotalLength = htons(sizeof(IPV4_HEADER) + sizeof(UDP_HEADER) + (uint16_t)Buffer->Length);
+        IPv4->Identification = 0;
+        IPv4->FlagsAndFragmentOffset = 0;
+        IPv4->TimeToLive = IP_DEFAULT_HOP_LIMIT;
+        IPv4->Protocol = IPPROTO_UDP;
+        IPv4->HeaderChecksum = 0;
+        CxPlatCopyMemory(IPv4->Destination, &RemoteAddress->Ipv4.sin_addr, sizeof(RemoteAddress->Ipv4.sin_addr));
+        CxPlatCopyMemory(IPv4->Source, &LocalAddress->Ipv4.sin_addr, sizeof(LocalAddress->Ipv4.sin_addr));
+        EthType = ETHERNET_TYPE_IPV4;
+        Ethernet = (ETHERNET_HEADER*)(((uint8_t*)IPv4) - sizeof(ETHERNET_HEADER));
+        IpHeaderLen = sizeof(IPV4_HEADER);
+    } else {
+        IPv6 = (IPV6_HEADER*)(((uint8_t*)UDP) - sizeof(IPV6_HEADER));
+        //
+        // IPv6 Version, Traffic Class, ECN Field and Flow Label fields in host
+        // byte order.
+        // TODO: support big-endian systems. Not quite sure if it's needed since QUIC core code
+        // also doesn't support big-endian systems due to the use of CxPlatByteSwapXxx which swaps
+        // bytes even when there's no need.
+        //
+        union {
+            struct {
+                uint32_t Flow : 20;
+                uint32_t EcnField : 2;
+                uint32_t Class : 6;
+                uint32_t Version : 4; // Most significant bits.
+            };
+            uint32_t Value;
+        } VersionClassEcnFlow = {0};
+
+        uint32_t FlowId = (uint32_t)(uintptr_t)Socket;
+        VersionClassEcnFlow.Version = IPV6_VERSION;
+        VersionClassEcnFlow.Class = 0;
+        VersionClassEcnFlow.EcnField = 0; // Not ECN capable currently.
+        VersionClassEcnFlow.Flow = FlowId;
+        if ((FlowId != 0) && (VersionClassEcnFlow.Flow == 0)) {
+            //
+            // If LSBs (20 bits) of Flow are all 0s, use its MSBs.
+            //
+            VersionClassEcnFlow.Flow = (FlowId >> 12);
+        }
```

Comment:
This is interesting, but is it strictly necessary?

Reply:
You meant the flow field is necessary or the part that moves bits around? Flow ID is not strictly necessary. This field is for routers and middlebox to do QoS or load distribution. It's recommended (https://datatracker.ietf.org/doc/html/rfc6437) to have one per flow but leaving to zero works too.

Reply:
I just mean, in general converting the socket pointer to a flow id. It requires an extra branch in the code that has a limited perf hit, for little or no real gain.

Reply:
Ok, I can try to skip the branch.